proposals/0000-term-lookup-rules.md

```diff
@@ -0,0 +1,199 @@
+  ```haskell
+  sizeOf :: forall a. Storable a => Int
+
+  sizeOf @MyType
+  ```
+  But that argument is implicit and not required. If the user forgets to specify it, GHC will try to infer it which will lead to bad error messages like "could not deduce" that does not really hint to what the problem is and how to fix it.
+
+These problems could be solved by changing the identifier lookup rules in a non-intrusive way: whenever we lookup an identifier at the term level we can first search for it in the data namespace, and if it wasn't found there, we could search in other namespaces as a fall-back. To keep compatibility with modules that use punning, a new syntax could be introduced allowing to import only terms or only types from a module.
+
+## Proposed Change Specification
+
+* The name lookup rules are extended: when looking up term-level identifier fails, look for a type-level identifier as a fall-back.
+* To disambiguate the namespaces of identifiers from modules that use punning, the `-XExplicitNamespaces` extension is extended with new syntax:
+  ```
+  impdecl   -> import [qualified] modid [as modid] [impspec]
+             |
+  ```
+  is changed to
+  ```
+  impdecl   -> import [qualified] modid [namespace] [as modid] [impspec]
+             |
+
+  namespace -> data
+             | type
+  ```
+   * With `data` specified in the import, only identifiers belonging to the data namespace will be brought into the scope.
+   * With `type` specified in the import, only identifiers belonging to the type namespace will be brought into the scope.
+* A new module `Data.Type` is introduced which defines unambiguous type definitions for built-in syntax constructors: `List`, `Tuple` and `Unit`:
+  ```haskell
+  type List = []
+  type Unit = ()
+  type Tuple2 = (,)
+  type Tuple3 = (,,)
+  type Tuple4 = (,,,)
+  {- ... -}
+
+  type family Tuple (xs :: [Type]) :: Type where
+    Tuple [] = Unit
+    Tuple [a] = GHC.Tuple.Unit a
+    Tuple [a,b] = Tuple2 a b
+    Tuple [a,b,c] = Tuple3 a b c
+    {- ... -}
+  ```
+* The `[a]` syntax means `[] a` when the `[]` type constructor is in scope, and a singleton list otherwise.
+* The `(a,b)` syntax means `(,) a b`, where `(,)` is either the type constructor or the data constructor according to the name lookup rules in the given context. This also applies to tuples of other arities.
+* `[]` and `()`, `(,)`, `(,,)`, ..., type constructors are no longer visible by default and instead are exported in a new module `Data.BuiltInTypes` which is imported by default like `Prelude`. `-XNoImplicitBuiltInTypes` is introduced to change this behavior.
+* Deprecate the `'` syntax of `-XDataKinds`, reserving this syntax for Template Haskell name quotation.
+* Deprecate the `''` syntax in Template Haskell.
+* Introduce a new warning, `-Wpun-declaration`, which warns the user when a declaration introduces a type constructor, type variable or a data constructor that clashes with an identifier in another namespace.
```

Comment:
What about a normal variable that clashes? For example:

```hs
foo :: forall x. x -> x
foo = bar
  where bar x = x
```

Reply:
Yes, this should still cause the warning. I've updated the proposal to encapture this scenario too and since this is not exactly a declaration I changed the warning's name to `-Wpun-bindings`, since I believe that is more fitting.